### PR TITLE
feat: add linux-riscv64 build

### LINKED ISSUE #7582: Provide riscv64 builds
### Contribution

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

As a developer, I want to be able to run pnpm on my risc-v system.
Additionally, as a devops person, I also want to be able to build a system built with riscv on my servers.

### Describe the solution you'd like

I was wondering if releasing builds for the riscv64 architecture is on the roadmap, and if not, if it could be taken into consideration.

RISC-V is an open standard instruction set architecture, which is under rapid development. Distributions including AOSP, Arch Linux, Debian, Gentoo Linux, OpenEuler, etc are also actively doing porting work for this architecture.

This repo tracks lots of this bringup work: https://github.com/carlosedp/riscv-bringup

### Describe the drawbacks of your solution

As it is still in the early days of this architecture, one drawback is that this work might not benefit all too many people yet.
As a counter-argument here, this is a bit of a chicken and egg situation.

### Describe alternatives you've considered

I really tried to find information on building pnpm from source, but couldnt find any. Perhaps creating this could be another solution.

### PR DESCRIPTION
**Pull request by velonica0**
Adds `linux-riscv64` to the Rust CLI's release targets, so pnpm 12 has a native binary on RISC-V. Closes pnpm/pnpm#7582.

Three places name a target, and this adds the same one to each:

- `release.yml` builds `riscv64gc-unknown-linux-gnu` with `cross`, on the existing `ubuntu-2404` runner that already produces both aarch64 targets. No new runner and no riscv64 hardware.
- `generate-packages.mjs` emits `@pnpm/exe.linux-riscv64` and lists it in the wrapper's `optionalDependencies`.
- `native-binary.mjs` resolves that package at install time.

No Rust or CLI logic changes.

### Relation to #8779

pnpm/pnpm#8779 added riscv64 in November 2024 and was reverted the next day in cebe493 (no PR, no stated reason). That version depended on Vercel `pkg` fetching a prebuilt `node20-linux-riscv64`, which `pkg-fetch` does not publish, so the target could only fail once a release actually ran. Since `release.yml` runs on tags and `workflow_dispatch` only, nothing in that PR could have failed in review.

The Rust CLI has no such dependency: `cross` supports `riscv64gc-unknown-linux-gnu` (`cpp`, `dylib`, `std`, `run`) and the binary is compiled rather than bundled around a prebuilt Node.

`pack_app.rs` still lists targets for `pnpm pack-app`, which embeds a Node binary via Node SEA. That constraint is unchanged, so this PR deliberately leaves it alone.

### Verification

The release workflow does not run on pull requests, so CI here will not build this target. Verified locally instead, against this branch. `cross` needs Docker, which was not available on the machine I tested on, so this is the same Cargo invocation `cross-build.sh` wraps, run against a native `riscv64-linux-gnu` toolchain. The sysroot and glibc therefore differ from what `cross` would use:

```
cargo build --locked -p pnpm-cli --bin pnpm --release --target riscv64gc-unknown-linux-gnu
  -> Finished `release` profile [optimized] in 9m 34s
  -> ELF 64-bit LSB pie executable, UCB RISC-V, RVC, double-float ABI
```

Under `qemu-riscv64`:

```
pnpm --version            -> 12.3.2
pnpm view is-odd version  -> 3.0.1     (live HTTPS to the registry)
```

The registry query matters because it exercises rustls and `aws-lc-rs` on riscv64. aws-lc-rs lists `riscv64gc-unknown-linux-gnu` as Build and Tests supported, and non-FIPS `aws-lc-sys` needs no CMake, bindgen, or Go.

Also checked `generate-packages.mjs` end to end in an isolated tree with the real binary staged. It produces `{"os":["linux"],"cpu":["riscv64"],"libc":["glibc"]}`, matching `@pnpm/exe.linux-arm64`'s shape, and `cpu: ["riscv64"]` matches what `@pnpm/linux-riscv64@9.14.0` published. Running `install.js` with `process.arch` forced to `riscv64` replaces the placeholder bin with the native binary at mode 755, which then runs.

A `workflow_dispatch` run on this branch before merge would confirm it in your CI rather than at tag time, which is where the 2024 attempt came apart.

## Squash Commit Body

```text
Build the Rust CLI for riscv64gc-unknown-linux-gnu and publish it as
@pnpm/exe.linux-riscv64, so installing pnpm on a riscv64 host resolves a
native binary instead of finding no prebuilt binary for the platform.

The release matrix builds the target with cross on the existing x86_64 Linux
runner, the same way it already builds both aarch64 targets, so no riscv64
hardware is involved.

PLATFORMS.linux gets a plain string specifier rather than a glibc/musl pair.
Only a glibc build is released, and the libc ordering maps over the pair's
keys, so a pair with a missing musl entry would yield an undefined specifier.

```
## Checklist

- [ ] I checked the referenced issue and verified that none of the PRs
  already linked to it solves it.
- [ ] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting.
- [ ] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Keep it short and written for pnpm users — it becomes a release note.
- [ ] Added or updated tests.
- [ ] Updated the documentation if needed.

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/14529"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1791100711&installation_model_id=426870&pr_number=14529&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F14529&signature=a9191ab909653dc65938ff46bf6fd7dc8b61820593567d647d1722a48cdfea8c"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

- **New Features**
  - Added support for running pnpm on glibc-based Linux systems with RISC-V 64-bit processors.
  - Added dedicated Linux RISC-V 64-bit native packages for improved platform compatibility.
  - Release builds now include the Linux RISC-V 64-bit target.
  - Added platform detection to select compatible native binaries and avoid unsupported libc combinations.
  - Improved fallback ordering on systems supporting multiple libc variants.
  - Updated supported-platform documentation to include Linux RISC-V 64-bit.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -595,6 +595,10 @@ jobs:
             target: aarch64-unknown-linux-musl
             code-target: linux-arm64-musl
 
+          - os: blacksmith-8vcpu-ubuntu-2404
+            target: riscv64gc-unknown-linux-gnu
+            code-target: linux-riscv64
```
[the remark this answers is not in the document]
Comment (velonica0) [reply .github/workflows/release.yml:600]: This is already in the commit the review ran on.

**Review by zkochan — APPROVED** [review]

**Comment by zkochan** [comment]
Checked both Qodo rule violations against the repo. Neither needs a change.

### 1. "Addon users miss the new platform" (rule 713441, changeset omits `@pnpm/napi`)

Not applicable here. `@pnpm/napi` is in a `versioning.fixed` group with `pacquet` in `pnpm-workspace.yaml`:

```yaml
fixed:
  - [ 'pacquet', '@pnpm/napi' ]
```

so it always releases at `pacquet`'s version and does not need its own changeset entry. `AGENTS.md` states the rule directly: "`@pnpm/napi` is a `versioning.fixed` group with it and bumps with it automatically."

`pnpm change status` on this branch confirms the pending plan already covers it:

```
@pnpm/napi: 12.3.4 -> 12.4.0 (minor, via fixed+intent)
pacquet:    12.3.4 -> 12.4.0 (minor, via fixed+intent)
```

Adding `@pnpm/napi` to the changeset would be redundant. The cited precedent, pnpm/pnpm#13492, concerns a published package outside a fixed group.

### 2. "Maintainers misread Linux host labels" (rule 713430, `hostTarget` JSDoc)

Declining, on two grounds.

The module-scoped capture of `platform` and `arch` is a property of the whole module, not of `hostTarget` — `getBinCandidates` and `resolveInstalledBinary` read the same two constants. `AGENTS.md` asks that a shared rationale be documented once at its common home rather than re-derived at each site, so attaching it to the one new helper would put it in the wrong place. It is already recorded where it actually bites: the test that has to re-import the module after faking the host.

The unknown-libc edge case has no reachable consequence. `hostTarget()` is only printed when `getBinCandidates()` returned empty, which happens in exactly two situations: an unsupported platform/arch, or a musl-probed riscv64 host. In the riscv64 case the probe succeeded by definition, so the label is accurate; in the unsupported-arch case no libc is supported, so the suffix carries no meaning. I enumerated all 60 platform/arch/libc combinations against both `main` and this branch to check this, which also confirmed that exactly one host's resolution changes:

```
DIFFERS linux/riscv64/glibc
  main: []
  PR:   ["@pnpm/exe.linux-riscv64/pnpm"]

checked 60 host combinations, 1 differ
```

Every host that installs today, under either libc and under a failed libc probe, gets an identical candidate list in the same order.

---
Written by an agent (Claude Code, claude-opus-5).
